File: behavioral/typescript/visitor_structured/shopping_cart.ts
import { Visitor, Item } from './interfaces';

export class Book implements Item {
  private title: string;
  private author: string;
  private price: number;
  
  constructor(title: string, author: string, price: number) {
    this.title = title;
    this.author = author;
    this.price = price;
  }
  
  accept(visitor: Visitor): void {
    visitor.visitBook(this);
  }
  
  getPrice(): number {
    return this.price;
  }
  
  getTitle(): string { return this.title; }
  getAuthor(): string { return this.author; }
}

export class Fruit implements Item {
  private name: string;
  private price: number;
  private weight: number;
  
  constructor(name: string, price: number, weight: number) {
    this.name = name;
    this.price = price;
    this.weight = weight;
  }
  
  accept(visitor: Visitor): void {
    visitor.visitFruit(this);
  }
  
  getPrice(): number {
    return this.price;
  }
  
  getName(): string { return this.name; }
  getWeight(): number { return this.weight; }
}

export class Electronic implements Item {
  private model: string;
  private brand: string;
  private price: number;
  
  constructor(brand: string, model: string, price: number) {
    this.brand = brand;
    this.model = model;
    this.price = price;
  }
  
  accept(visitor: Visitor): void {
    visitor.visitElectronic(this);
  }
  
  getPrice(): number {
    return this.price;
  }
  
  getModel(): string { return this.model; }
  getBrand(): string { return this.brand; }
}

export class PriceCalculator implements Visitor {
  private totalPrice: number = 0;
  
  visitBook(book: any): void {
    this.totalPrice += book.getPrice();
    console.log(`Book: ${book.getTitle()} - $${book.getPrice()}`);
  }
  
  visitFruit(fruit: any): void {
    this.totalPrice += fruit.getPrice();
    console.log(`Fruit: ${fruit.getName()} - $${fruit.getPrice()}`);
  }
  
  visitElectronic(electronic: any): void {
    this.totalPrice += electronic.getPrice();
    console.log(`Electronic: ${electronic.getBrand()} ${electronic.getModel()} - $${electronic.getPrice()}`);
  }
  
  getTotalPrice(): number {
    return this.totalPrice;
  }
}

export class DescriptionGenerator implements Visitor {
  visitBook(book: any): void {
    console.log(`Book: "${book.getTitle()}" by ${book.getAuthor()} - A great read!`);
  }
  
  visitFruit(fruit: any): void {
    console.log(`Fruit: ${fruit.getName()} (${fruit.getWeight()}kg) - Fresh and healthy!`);
  }
  
  visitElectronic(electronic: any): void {
    console.log(`Electronic: ${electronic.getBrand()} ${electronic.getModel()} - High-tech gadget!`);
  }
}

export class TaxCalculator implements Visitor {
  private totalTax: number = 0;
  
  visitBook(book: any): void {
    const tax = book.getPrice() * 0.05;
    this.totalTax += tax;
    console.log(`Book tax: $${tax}`);
  }
  
  visitFruit(fruit: any): void {
    const tax = fruit.getPrice() * 0.02;
    this.totalTax += tax;
    console.log(`Fruit tax: $${tax}`);
  }
  
  visitElectronic(electronic: any): void {
    const tax = electronic.getPrice() * 0.10;
    this.totalTax += tax;
    console.log(`Electronic tax: $${tax}`);
  }
  
  getTotalTax(): number {
    return this.totalTax;
  }
}

export class ShoppingCart {
  private items: Item[] = [];
  
  addItem(item: Item): void {
    this.items.push(item);
  }
  
  acceptVisitor(visitor: Visitor): void {
    for (const item of this.items) {
      item.accept(visitor);
    }
  }
}

File: behavioral/dart/visitor_structured/shopping_cart.dart
import 'interfaces.dart';

class Book extends Item {
  final String _title;
  final String _author;
  final double _price;
  
  Book(this._title, this._author, this._price);
  
  @override
  void accept(Visitor visitor) {
    visitor.visitBook(this);
  }
  
  @override
  double getPrice() => _price;
  
  String get title => _title;
  String get author => _author;
}

class Fruit extends Item {
  final String _name;
  final double _price;
  final double _weight;
  
  Fruit(this._name, this._price, this._weight);
  
  @override
  void accept(Visitor visitor) {
    visitor.visitFruit(this);
  }
  
  @override
  double getPrice() => _price;
  
  String get name => _name;
  double get weight => _weight;
}

class Electronic extends Item {
  final String _model;
  final String _brand;
  final double _price;
  
  Electronic(this._brand, this._model, this._price);
  
  @override
  void accept(Visitor visitor) {
    visitor.visitElectronic(this);
  }
  
  @override
  double getPrice() => _price;
  
  String get model => _model;
  String get brand => _brand;
}

class PriceCalculator implements Visitor {
  double _totalPrice = 0;
  
  @override
  void visitBook(Book book) {
    _totalPrice += book.getPrice();
    print('Book: ${book.title} - \$${book.getPrice()}');
  }
  
  @override
  void visitFruit(Fruit fruit) {
    _totalPrice += fruit.getPrice();
    print('Fruit: ${fruit.name} - \$${fruit.getPrice()}');
  }
  
  @override
  void visitElectronic(Electronic electronic) {
    _totalPrice += electronic.getPrice();
    print('Electronic: ${electronic.brand} ${electronic.model} - \$${electronic.getPrice()}');
  }
  
  double get totalPrice => _totalPrice;
}

class DescriptionGenerator implements Visitor {
  @override
  void visitBook(Book book) {
    print('Book: "${book.title}" by ${book.author} - A great read!');
  }
  
  @override
  void visitFruit(Fruit fruit) {
    print('Fruit: ${fruit.name} (${fruit.weight}kg) - Fresh and healthy!');
  }
  
  @override
  void visitElectronic(Electronic electronic) {
    print('Electronic: ${electronic.brand} ${electronic.model} - High-tech gadget!');
  }
}

class TaxCalculator implements Visitor {
  double _totalTax = 0;
  
  @override
  void visitBook(Book book) {
    final tax = book.getPrice() * 0.05;
    _totalTax += tax;
    print('Book tax: \$${tax}');
  }
  
  @override
  void visitFruit(Fruit fruit) {
    final tax = fruit.getPrice() * 0.02;
    _totalTax += tax;
    print('Fruit tax: \$${tax}');
  }
  
  @override
  void visitElectronic(Electronic electronic) {
    final tax = electronic.getPrice() * 0.10;
    _totalTax += tax;
    print('Electronic tax: \$${tax}');
  }
  
  double get totalTax => _totalTax;
}

class ShoppingCart {
  final List<Item> _items = [];
  
  void addItem(Item item) {
    _items.add(item);
  }
  
  void acceptVisitor(Visitor visitor) {
    for (final item in _items) {
      item.accept(visitor);
    }
  }
}
